src/ethereum/frontier/vm/instructions/bitwise.py

```diff
@@ -0,0 +1,148 @@
+    result back on the stack.
+
+    Parameters
+    ----------
+    evm :
+        The current EVM frame.
+
+    Raises
+    ------
+    StackUnderflowError
+        If `len(stack)` is less than `2`.
+    OutOfGasError
+        If `evm.gas_left` is less than `GAS_VERY_LOW`.
+    """
+    evm.gas_left = subtract_gas(evm.gas_left, GAS_VERY_LOW)
+    x = pop(evm.stack)
+    y = pop(evm.stack)
+    push(evm.stack, x & y)
+
+
+def bitwise_or(evm: Evm) -> None:
+    """
+    Bitwise OR operation of the top 2 elements of the stack. Pushes the
+    result back on the stack.
+
+    Parameters
+    ----------
+    evm :
+        The current EVM frame.
+
+    Raises
+    ------
+    StackUnderflowError
+        If `len(stack)` is less than `2`.
+    OutOfGasError
+        If `evm.gas_left` is less than `GAS_VERY_LOW`.
+    """
+    evm.gas_left = subtract_gas(evm.gas_left, GAS_VERY_LOW)
+    x = pop(evm.stack)
+    y = pop(evm.stack)
+    push(evm.stack, x | y)
+
+
+def bitwise_xor(evm: Evm) -> None:
+    """
+    Bitwise XOR operation of the top 2 elements of the stack. Pushes the
+    result back on the stack.
+
+    Parameters
+    ----------
+    evm :
+        The current EVM frame.
+
+    Raises
+    ------
+    StackUnderflowError
+        If `len(stack)` is less than `2`.
+    OutOfGasError
+        If `evm.gas_left` is less than `GAS_VERY_LOW`.
+    """
+    evm.gas_left = subtract_gas(evm.gas_left, GAS_VERY_LOW)
+    x = pop(evm.stack)
+    y = pop(evm.stack)
+    push(evm.stack, x ^ y)
+
+
+def bitwise_not(evm: Evm) -> None:
+    """
+    Bitwise NOT operation of the top element of the stack. Pushes the
+    result back on the stack.
+
+    Parameters
+    ----------
+    evm :
+        The current EVM frame.
+
+    Raises
+    ------
+    StackUnderflowError
+        If `len(stack)` is less than `1`.
+    OutOfGasError
+        If `evm.gas_left` is less than `GAS_VERY_LOW`.
+    """
+    evm.gas_left = subtract_gas(evm.gas_left, GAS_VERY_LOW)
+    x = pop(evm.stack)
+    push(evm.stack, ~x)
+
+
+def get_byte(evm: Evm) -> None:
+    """
+    For a word (defined by next top element of the stack), retrieve the
+    Nth byte (0-indexed and defined by top element of stack) from the left
+    to right (MSB to LSB).
+
+    Parameters
+    ----------
+    evm :
+        The current EVM frame.
+
+    Raises
+    ------
+    StackUnderflowError
+        If `len(stack)` is less than `2`.
+    OutOfGasError
+        If `evm.gas_left` is less than `GAS_VERY_LOW`.
+    """
+    evm.gas_left = subtract_gas(evm.gas_left, GAS_VERY_LOW)
+    # 0-indexed from MSB to LSB (left to right in "Big Endian" representation)
+    byte_index = pop(evm.stack)
+    word = pop(evm.stack)
+
+    if byte_index >= 32:
+        result = U256(0)
+    else:
+        extra_bytes_to_left = byte_index
+        extra_bytes_to_right = 31 - byte_index
+        # Remove the extra bytes in the right
+        word = word >> (extra_bytes_to_right * 8)
+        if extra_bytes_to_left != 0:
```

Comment:
I think I would prefer to just shift right and always `&` out the upper bits, even if it's slightly less efficient.

Reply:
I initially did it that way, but later felt that this optimization is pretty readable and highly explicit. Can you please let me know why you prefer to always perform the `&` operation? Thanks.

Reply:
First, I don't think we should be optimizing at all, especially without data. Second, I would be incredibly surprised if a branch is more efficient than just a shift and mask.

I did a little informal benchmarking, just because I was curious. I wouldn't consider these results as rigorous.

Here's my test harness:

```python
from ethereum.frontier.vm import Evm
from ethereum.frontier.vm.instructions.bitwise import get_byte
import timeit

def test():
    evm = Evm(pc=0, stack=[0xAABBCC, 30], memory=[], code=[], gas_left=10000000, current=[], caller=[], data=[], value=0, depth=0, env=None, refund_counter=0, running=True)
    get_byte(evm)
    return evm.stack[0]

timeit.timeit("test()", globals=locals())
```

With the branch, it took `2.8620279809983913` seconds. Without the branch, it took `2.770114379003644`.

Reply:
Thanks for this benchmark Sam. Learned from it. I'll remove the branch.